content/blog/2023-08-03-parsing-ssh-config.markdown

```diff
@@ -0,0 +1,229 @@
+the original window.
+
+In this block, there is a conditional statement on whether the `ProxyJump` field
+is present in the config. This is a good place to break on and see what the
+computed config looks like. If it doesn't find a proxy then of course it's going
+to run everything on the host.
+
+And indeed, it doesn't think there is a proxy. This is progress, but why does
+the extension's view of the config not match up with what SSH does? After all,
+invoking SSH directly connects properly. Tracing back the source of the config
+in the extension, it ultimately comes from manually reading in and parsing the
+SSH config. When resolving the host argument it manually computes the config as
+per [`ssh_config(5)`](https://man7.org/linux/man-pages/man5/ssh_config.5.html).
+Yet somewhere it makes a mistake, because it doesn't include the `ProxyJump`
+field.
+
+
+## Parsing SSH config
+
+To get to the bottom of this, we need to know the rules behind parsing SSH
+configs. The `ssh_config(5)` manpage does a pretty decent job of explaining
+this, but I'm going to go over the relevant information here. I reckon most
+people have a vague idea of how it works, and can write enough to meet their
+needs, but have never looked deeper into the actual rules behind how SSH parses
+the config.
+
+1. For starters, the config is parsed line by line. Leading whitespace (i.e.,
+   indentation) is ignored. So, while indentation makes it look like you are
+   configuring properties for a particular host, this isn't quite correct.
+   Instead, the `Host` and `Match` lines are special statements that enable or
+   disable all subsequent lines until the next `Host` or `Match`.
+
+    There is no backtracking; previous conditions and lines are not re-evaluated
+    after learning more about the config later on.
+
+2. When a config line is seen, and is active thanks to the most recent `Host` or
+   `Match` succeeding, its value is selected if it is the first of that config
+   to be selected. So the earliest place a value is set takes priority; this may
+   be a little counterintuitive if you are used to having the latest value be
+   picked, such as enable/disable command line flags tend to work.
+
+3. When `HostName` is set, it replaces the value used in `Host` matches and the
+   `host` value in `Match` matches.
+
+4. The last behaviour of interest is the `Match final` rule. There are several
+   conditions a `Match` statement can have, and the `final` rule says make this
+   active on the final pass over the config.
+
+Wait, final pass? Multiple passes? Yes. If `final` is a condition on a `Match`,
+SSH will do another pass over the config, following all the rules above. Except
+this time all the configs we read on the first pass are still active (and can't
+be changed). But all the `Host` and `Matches` are re-evaluated, allowing other
+configs to potentially be set. I guess that means rule (1) ought to have a big
```

Comment:
mention explicitly that the configs that can be affected aren't just the rules that are contained within the `Match final` as you go on to elaborate in the example below